validationparser/interceptor.go

```diff
@@ -0,0 +1,46 @@
+package validationparser
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type validator interface {
+	Validate() error
+}
+
+// UnaryServerInterceptor returns grpc.UnaryServerInterceptor
+// that should be used as a middleware that parser through the protoc-gen-validate middleware and returns a
+// user friendly error.
+func UnaryServerInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		if v, ok := req.(validator); ok {
+			if err := v.Validate(); err != nil {
+				// Check what type of error we received
+				typeOfError := reflect.TypeOf(err)
+				valueOfError := reflect.ValueOf(err)
+				// If the error is a type of validation error then parse
+				if strings.Contains(typeOfError.String(), "ValidationError") && valueOfError.NumField() == 4 {
+					// Get the cause of error from the error struct
+					cause := valueOfError.FieldByName("Cause")
```

Comment:
You should probably check that the error is a struct before doing Numfield() (even if the name contains ValidationError), https://stackoverflow.com/questions/42065226/how-to-determine-if-type-is-a-struct-in-golang
And also check that cause is a valid field https://golang.org/pkg/reflect/#Value.IsValid before doing TypeOf or ValueOf to prevent panics. 
I know I put you on this path, but I'd still love it if somebody had a better solution.

Reply:
and add unit tests to verify all this!